libres/lib/enkf/enkf_main.cpp

```diff
@@ -743,6 +737,337 @@ static FILE *enkf_main_log_step_list(const char *log_path,
+
+        if (!(local_dataset_get_size(dataset)))
+            continue;
+
+        local_obsdata_type *local_obsdata =
+            local_ministep_get_obsdata(ministep);
+        const auto &unscaled_keys = local_dataset_unscaled_keys(dataset);
+
+        if (unscaled_keys.size() == 0)
+            continue;
+
+        // Part 1: Parameters which do not have row scaling attached.
+        enkf_main_serialize_dataset(ensemble_config, dataset, last_step,
+                                    use_count, tp, serialize_info);
+
+        matrix_type *A_new = matrix_alloc_copy(A);
+        parameters[std::string(dataset_name)] = A_new;
+    }
+    serialize_info_free(serialize_info);
+    matrix_free(A);
+    thread_pool_free(tp);
+    return parameters;
+}
+
+static void enkf_main_save_parameters_from_ministep(
+    enkf_fs_type *target_fs, ensemble_config_type *ensemble_config,
+    int_vector_type *iens_active_index, int last_step, run_mode_type run_mode,
+    enkf_state_type **ensemble, hash_type *use_count,
+    const local_ministep_type *ministep,
+    std::unordered_map<std::string, matrix_type *> parameters) {
+
+    int cpu_threads = 4;
+    thread_pool_type *tp = thread_pool_alloc(cpu_threads, false);
+
+    for (auto &[dataset_name, dataset] :
+         local_ministep_get_datasets(ministep)) {
+
+        if (!(local_dataset_get_size(dataset)))
+            continue;
+
+        local_obsdata_type *local_obsdata =
+            local_ministep_get_obsdata(ministep);
+        const auto &unscaled_keys = local_dataset_unscaled_keys(dataset);
+
+        if (unscaled_keys.size() == 0)
+            continue;
+
+        matrix_type *A = parameters[std::string(dataset_name)];
+        serialize_info_type *serialize_info = serialize_info_alloc(
+            target_fs, //src_fs - we have already copied the parameters from the src_fs to the target_fs
+            target_fs, ensemble_config, iens_active_index, 0, ensemble,
+            run_mode, last_step, matrix_alloc_copy(A), cpu_threads);
+
+        enkf_main_serialize_dataset(ensemble_config, dataset, last_step,
+                                    use_count, tp, serialize_info);
+        for (int i = 0; i < cpu_threads; i++) {
+            serialize_info[i].A = matrix_alloc_copy(A);
+        }
+
+        enkf_main_deserialize_dataset(ensemble_config, dataset, serialize_info,
+                                      tp);
+        serialize_info_free(serialize_info);
+    }
+    thread_pool_free(tp);
+}
+
+static std::unordered_map<
+    std::string,
+    std::vector<std::pair<matrix_type *, const row_scaling_type *>>>
+enkf_main_load_row_scaling_parameters(
+    enkf_fs_type *target_fs, ensemble_config_type *ensemble_config,
+    int_vector_type *iens_active_index, int last_step, run_mode_type run_mode,
+    meas_data_type *forecast, enkf_state_type **ensemble, hash_type *use_count,
+    obs_data_type *obs_data, const local_ministep_type *ministep) {
+
+    int matrix_start_size = 250000;
```

Comment:
`matrix_start_size` is set to 250000 in this function but also in `enkf_main_load_parameters_from_ministep`.
Perhaps add `matrix_start_size` as an input parameter and set it in `enkf_main_update__`?

Reply:
these functions already have waaayyy to many parameters. I think we should figure out how we can refactor this before we introduce new ones.